programs/ssl/ssl_client2.c

```diff
@@ -2382,15 +2433,22 @@ int main( int argc, char *argv[] )
     }
     else
 #endif /* MBEDTLS_USE_PSA_CRYPTO */
-    if( psk_len > 0 )
+    if( psk_info.psk_len > 0 )
     {
-        ret = mbedtls_ssl_conf_psk( &conf, psk, psk_len,
-                             (const unsigned char *) opt.psk_identity,
-                             strlen( opt.psk_identity ) );
-        if( ret != 0 )
+        if (opt.psk_callback == 0) {
+            ret = mbedtls_ssl_conf_psk( &conf, psk_info.psk, psk_info.psk_len,
+                                 (const unsigned char *) opt.psk_identity,
+                                 strlen( opt.psk_identity ) );
+            if( ret != 0 )
+            {
+                mbedtls_printf( " failed\n  ! mbedtls_ssl_conf_psk returned %d\n\n", ret );
+                goto exit;
+            }
+        }
+        else
         {
-            mbedtls_printf( " failed\n  ! mbedtls_ssl_conf_psk returned %d\n\n", ret );
-            goto exit;
+            psk_info.psk_server_identity = opt.psk_server_identity;
+            mbedtls_ssl_conf_psk_cb(&conf, &psk_cb, &psk_info);
```

Comment:
Here I think we also want to call `mbedtls_ssl_conf_psk()` with a dummy buffer for the key, and the correct identity. That way we'll have the identity set, and won't need to set it in the callback (while the key set from the callback will overwrite the one set here).